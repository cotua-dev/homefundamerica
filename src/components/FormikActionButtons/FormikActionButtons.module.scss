div.actionButtonsWrapper {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin: 40px 0;

    @media screen and (max-width: 992px) {
        margin: 10px 0;
    }

    @media screen and (max-width: 900px) {
        justify-content: center;
        flex-direction: column;
        margin: 10px 0;
    }

    button.actionButton {
        width: 200px;
        height: 50px;
        font-family: 'Inter', Helvetica, sans-serif;
        font-size: 14px;
        font-weight: 700;
        outline: none;
        border: 1px solid #1F72C1;
        border-radius: 30px;
        padding: 10px 20px;
        margin: 0 20px;
        cursor: pointer;
        background-color: #E8F0F8;
        color: #1F72C1;
        transition: 0.3s;

        &:hover {
            background-color: #1F72C1;
            color: #FFFFFF;
            transform: translateY(-3px);
            box-shadow: 0 14px 28px rgba(0, 0, 0, .25), 0 10px 10px rgba(0, 0, 0, .22);
        }

        &:focus {
            outline: none;
        }

        &:disabled {
            border: 1px solid #C41A2F;
            background-color: #F9E8EA;
            color: #C41A2F;
            cursor: initial;
        }

        @media screen and (max-width: 900px) {
            margin: 10px 0;
            padding: 5px;
        }

        .loadingSpinner {
            font-size: 24px;
        }
    }

    & button.actionButton:not(:first-child) {
        margin-left: 20px;

        @media screen and (max-width: 900px) {
            margin: 0;
        }
    }
}
