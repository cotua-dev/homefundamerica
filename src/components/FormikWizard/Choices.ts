import {
    FirstChoice,
    SecondChoice,
    ThirdChoice,
    FourthChoice,
} from "./FormikWizard.interfaces";

export const firstStepChoices: FirstChoice[] = [
    {
        key: 0,
        value: 'Purchase',
        image: '/purchase.svg',
    },
    {
        key: 1,
        value: 'Refinance',
        image: '/refinance.svg',
    },
];

export const secondStepChoices: SecondChoice[] = [
    {
        key: 2,
        value: 'Primary',
        image: '/primary.svg',
    },
    {
        key: 3,
        value: 'Vacation',
        image: '/vacation.svg',
    },
    {
        key: 4,
        value: 'Investment',
        image: '/investment.svg',
    },
];

export const thirdStepChoices: ThirdChoice[] = [
    {
        key: 5,
        value: 'Single Family',
        image: '/single-family.svg',
    },
    {
        key: 6,
        value: 'Condo',
        image: '/condo.svg',
    },
    {
        key: 7,
        value: 'Multi-Family',
        image: '/multi-family.svg',
    },
    {
        key: 8,
        value: 'Townhouse',
        image: '/townhouse.svg',
    },
];

export const fourthStepChoices: FourthChoice[] = [
    {
        key: 9,
        value: 'Low (619 or less)',
        image: '/1-star.svg',
    },
    {
        key: 10,
        value: 'Fair (620 - 679)',
        image: '/2-stars.svg',
    },
    {
        key: 11,
        value: 'Good (680 - 739)',
        image: '/3-stars.svg',
    },
    {
        key: 12,
        value: 'Excellent (740+)',
        image: '/4-stars.svg',
    },
];
