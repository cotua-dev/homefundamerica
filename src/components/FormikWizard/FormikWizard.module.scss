@import '../../styles/variables';

section.formikSection {
    width: 100%;
    height: 700px;
    background-color: #F8FAFC;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;

    @media screen and (max-width: 1450px) {
        height: 550px;
    }

    @media screen and (max-width: 992px) {
        height: 100%;
    }

    div.namesWrapper {
        display: flex;
        flex-direction: column;
    }

    div.step {
        display: flex;
        align-items: center;
        justify-content: center;
        flex-direction: column;

        input[type=number]::-webkit-outer-spin-button,
        input[type=number]::-webkit-inner-spin-button {
            appearance: none;
            -webkit-appearance: none;
            -moz-appearance: none;
            margin: 0;
        }

        input[type=number] {
            -moz-appearance: textfield;
        }

        input.field {
            font-family: 'Inter', Helvetica, sans-serif;
            font-size: 20px;
            margin-bottom: 20px;
            border: 1px solid #E0E0E0;
            border-radius: 10px;
            width: 80%;
            padding: 15px 10px;
        }

        h3.stepTitle {
            font-family: 'Inter', Helvetica, sans-serif;
            font-size: 36px;
            font-weight: 900;
            color: $primaryColor;
            line-height: 47px;
            margin: 20px 0;
            text-align: center;

            @media screen and (max-width: 900px) {
                font-size: 18px;
                line-height: 20px;
                margin: 10px 0;
            }
        }

        div.choices {
            display: flex;
            align-items: center;
            justify-content: center;
            flex-wrap: wrap;

            @media screen and (max-width: 900px) {
                flex-wrap: nowrap;
                flex-direction: column;
                width: 100%;
            }

            div.stepChoice {
                height: 350px;
                width: 280px;
                display: flex;
                align-items: center;
                justify-content: space-between;
                flex-direction: column;
                padding: 16px;
                margin: 15px;
                border: 1px solid $primaryColor;
                border-radius: 5px;
                cursor: pointer;
                box-shadow: 0 1px 3px rgba(0, 0, 0, .12), 0 1px 2px rgba(0, 0, 0, .24);
                transition: all 0.3s cubic-bezier(.25, .8, .25, 1);

                &:hover {
                    box-shadow: 0 14px 28px rgba(0, 0, 0, .25), 0 10px 10px rgba(0, 0, 0, .22);
                }

                @media screen and (max-width: 1450px) {
                    height: 200px;
                    width: 160px;
                }

                @media screen and (max-width: 900px) {
                    width: 280px;
                    height: 80px;
                    padding: 5px;
                    margin: 5px;
                    align-items: center;
                    flex-direction: row;
                    justify-content: flex-start;
                }

                div.choiceMark {
                    height: 26px;
                    width: 26px;
                    margin-right: auto;
                    margin-bottom: 20px;
                    border: 2px solid $primaryColor;
                    border-radius: 15px;
                    display: flex;
                    align-items: center;
                    justify-content: center;

                    @media screen and (max-width: 1450px) {
                        margin-bottom: 0;
                        height: 20px;
                        width: 20px;
                    }

                    @media screen and (max-width: 900px) {
                        height: 18px;
                        width: 18px;
                        margin-right: 0;
                    }

                    div.choiceMarkFiller {
                        height: 16px;
                        width: 16px;
                        background-color: $primaryColor;
                        border-radius: 8px;

                        @media screen and (max-width: 1450px) {
                            height: 12px;
                            width: 12px;
                        }

                        @media screen and (max-width: 500px) {
                            height: 12px;
                            width: 12px;
                        }
                    }
                }

                img.choiceImage {
                    height: auto;
                    width: 250px;
                    padding: 0 20px;

                    @media screen and (max-width: 1450px) {
                        width: 100px;
                    }

                    @media screen and (max-width: 900px) {
                        width: 75px;
                    }
                }

                h5.choiceTitle {
                    font-family: 'Inter', Helvetica, sans-serif;
                    font-size: 20px;
                    font-weight: 700;
                    color: $primaryColor;
                    margin: 0;

                    @media screen and (max-width: 1450px) {
                        font-size: 18px;
                    }

                    @media screen and (max-width: 900px) {
                        font-size: 16px;
                        margin: 0 auto;
                    }
                }
            }
        }
    }

    div.thankYouWrapper {
        display: flex;
        align-items: center;
        justify-content: center;
        text-align: center;
        flex-direction: column;

        strong.thanksTitle {
            width: 500px;
            font-size: 20px;
            color: $primaryColor;
            margin: 20px 0 10px 0;

            @media screen and (max-width: 600px) {
                width: 100%;
                margin: 20px;
            }
        }

        a.homePageLink {
            display: flex;
            align-items: center;
            justify-content: center;
            width: 150px;
            height: 25px;
            font-size: 18px;
            font-weight: 700;
            text-decoration: none;
            border: 1px solid $primaryColor;
            border-radius: 30px;
            padding: 10px 20px;
            margin: 10px 0 20px 0;
            background-color: #E8F0F8;
            color: $primaryColor;
            transition: 0.8s;

            &:hover {
                background-color: $primaryColor;
                color: $textColor;
            }
        }
    }

    div.errorMessageWrapper {
        display: flex;
        align-items: center;
        justify-content: center;
        flex-direction: column;
        margin: 40px 10px;

        strong.errorMessage {
            color: $primaryColor;
            font-size: 28px;
            margin-bottom: 10px;
            text-align: center;

            @media screen and (max-width: 950px) {
                font-size: 24px;
            }
        }

        button.resetButton {
            width: 200px;
            height: 50px;
            font-family: 'Inter', Helvetica, sans-serif;
            font-size: 14px;
            font-weight: 700;
            outline: none;
            border: 1px solid $primaryColor;
            border-radius: 30px;
            padding: 10px 20px;
            margin: 0 20px;
            cursor: pointer;
            background-color: #E8F0F8;
            color: $primaryColor;
            transition: 0.3s;

            &:hover {
                background-color: $primaryColor;
                color: $textColor;
                transform: translateY(-3px);
                box-shadow: 0 14px 28px rgba(0, 0, 0, .25), 0 10px 10px rgba(0, 0, 0, .22);
            }
        }
    }
}