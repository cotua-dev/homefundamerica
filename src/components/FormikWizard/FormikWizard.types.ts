export type PossibleChoices = 0 | 1 | 2 | 3;
