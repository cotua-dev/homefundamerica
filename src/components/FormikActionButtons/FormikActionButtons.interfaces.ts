import { FormikProps } from "formik";
import { FormikValues } from "../FormikWizard/FormikWizard.interfaces";

export interface Props
{
    canStep: boolean;
    stepNumber: number;
    totalSteps: number;
    formikProps: FormikProps<FormikValues>;
    onNext: () => void;
    onPrevious: () => void;
}
