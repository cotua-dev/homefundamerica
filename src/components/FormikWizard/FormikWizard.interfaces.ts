import { PossibleChoices } from './FormikWizard.types';

export interface State
{
    canStep: boolean;
    stepNumber: number;
    isSuccess: boolean;
    isError: boolean;
    errorMessage: string;
}

export interface PostResponseData
{
    Message: string;
}

export interface HubspotHeader
{
    Key: string;
    Value: string[];
}

export interface HubspotRequestMessage
{
    Content: {
        Headers: HubspotHeader[];
    };
    Headers: HubspotHeader[];
    Method: { Method: string };
    Properties: any;
    RequestUri: string;
    Version: HubspotVersion;
}

export interface HubspotVersion
{
    Build: number;
    Major: number;
    MajorRevision: number;
    Minor: number;
    MinorRevision: number;
    Revision: number;
}

export interface HubspotResponseJSON
{
    Content: {
        Headers: HubspotHeader[];
    };
    Headers: HubspotHeader[];
    IsSuccessStatusCode: boolean;
    ReasonPhrase: string;
    RequestMessage: HubspotRequestMessage;
    RequestUri: string;
    StatusCode: number;
    TrailingHeaders: HubspotHeader[];
    Version: HubspotVersion;
}

export interface InjectedField
{
    field: () => JSX.Element | undefined;
}

export interface Choices
{
    purchaseOrRefinance: PossibleChoices;
    propertyUse: PossibleChoices;
    propertyType: PossibleChoices;
    creditScore: PossibleChoices;
}

export interface FormikValues
{
    purchaseOrRefinance: null | Choices['purchaseOrRefinance'];
    propertyUse: null | Choices['propertyUse'];
    propertyType: null | Choices['propertyType'];
    credit: null | Choices['creditScore'];
    borrowAmount: null | number;
    loanAmount: null | number;
    propertyValue: null | number;
    propertyCity: null | string;
    firstName: null | string;
    lastName: null | string;
    email: null | string;
    phone: null | string;
}

export interface Field
{
    key: string;
    Element: JSX.Element;
}

export interface Choice
{
    key: number;
    image?: string;
    text: string;
}

export interface FirstChoice extends Choice
{
    value: FormikValues['purchaseOrRefinance'];
}

export interface SecondChoice extends Choice
{
    value: FormikValues['propertyUse'];
}

export interface ThirdChoice extends Choice
{
    value: FormikValues['propertyType'];
}

export interface FourthChoice extends Choice
{
    value: FormikValues['credit'];
}

export interface StepChoiceProps
{
    key: number;
    onClick: () => void;
    state: FormikValues;
}

export interface StepOneChoiceProps extends StepChoiceProps
{
    value: FirstChoice['value'];
}

export interface StepTwoChoiceProps extends StepChoiceProps
{
    value: SecondChoice['value'];
}

export interface StepThreeChoiceProps extends StepChoiceProps
{
    value: ThirdChoice['value'];
}

export interface StepFourChoiceProps extends StepChoiceProps
{
    value: FourthChoice['value'];
}

export interface ChoiceMarkFillerProps
{
    state: FormikValues;
}

export interface ChoiceOneMarkFillerProps extends ChoiceMarkFillerProps
{
    value: FirstChoice['value'];
}

export interface ChoiceTwoMarkFillerProps extends ChoiceMarkFillerProps
{
    value: SecondChoice['value'];
}

export interface ChoiceThreeMarkFillerProps extends ChoiceMarkFillerProps
{
    value: ThirdChoice['value'];
}

export interface ChoiceFourMarkFillerProps extends ChoiceMarkFillerProps
{
    value: FourthChoice['value'];
}
