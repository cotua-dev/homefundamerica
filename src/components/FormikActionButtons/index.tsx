import React, {
    FunctionComponent,
    ReactElement,
    useContext,
} from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';
import { Props } from './FormikActionButtons.interfaces';
import * as styles from './FormikActionButtons.module.scss';
import { FormikValues } from '../FormikWizard/FormikWizard.interfaces';
import {
    Interface,
    InterfaceType,
} from '../../contexts';

const FormikActionButtons: FunctionComponent<Props> = (props: Props): ReactElement =>
{
    const { isLoading } = useContext<InterfaceType>(Interface);

    /**
     * Check to see if formik field has a value
     * 
     * @param field Formik field name
     * @returns Boolean depending on presence of formik field value
     */
    const checkFormikValue = (field: keyof FormikValues): boolean =>
    {
        return props.formikProps.values[field] === null ? false : true;
    }

    /**
     * Check the borrow amount field value is
     * greater than or equal to 100000
     */
    // const checkBorrowAmount = (): boolean => Number(props.formikProps.values.borrowAmount) >= 100000 ? true : false;

    /**
     * Check if the property value field is
     * greater than or equal to 100000
     */
    // const checkLoanAmount = (): boolean => Number(props.formikProps.values.loanAmount) >= 100000 ? true : false;

    /**
     * Check purchaseOrRefinance field to update switch statement
     * 
     * @returns Result of check as boolean, null, or undefined values
     */
    const checkPurchaseOrRefinanceValue = (): boolean | null | undefined =>
    {
        const { purchaseOrRefinance } = props.formikProps.values;

        if (purchaseOrRefinance === null)
        {
            return null;
        } else {
            if (purchaseOrRefinance === 'Purchase')
            {
                return true;
            }

            if (purchaseOrRefinance === 'Refinance')
            {
                return false;
            }
        }
    }

    /**
     * Check to see if the next step is allowed
     * 
     * @returns Boolean depending on allowance
     */
    const allowNextStep = (): boolean =>
    {
        let allowance = false;
        const purchaseOrRefinance = checkPurchaseOrRefinanceValue();

        switch (props.stepNumber)
        {
            case 1:
                allowance = checkFormikValue('purchaseOrRefinance')
                break;
            case 2:
                allowance = checkFormikValue('propertyUse');
                break;
            case 3:
                allowance = checkFormikValue('propertyType');
                break
            case 4:
                allowance = checkFormikValue('credit');
                break;
        }

        if (typeof purchaseOrRefinance === 'boolean')
        {
            if (purchaseOrRefinance)
            {
                switch (props.stepNumber)
                {
                    case 5:
                        // allowance = checkBorrowAmount();
                        allowance = checkFormikValue('borrowAmount');
                        break;
                    case 6:
                        allowance = checkFormikValue('propertyCity');
                        break;
                    case 7:
                        allowance = checkFormikValue('firstName') && checkFormikValue('lastName');
                        break;
                    case 8:
                        allowance = checkFormikValue('email');
                        break;
                    case 9:
                        allowance = checkFormikValue('phone');
                        break;
                }
            } else
            {
                switch (props.stepNumber)
                {
                    case 5:
                        // allowance = checkLoanAmount();
                        allowance = checkFormikValue('loanAmount');
                        break;
                    case 6:
                        allowance = checkFormikValue('propertyValue');
                        break;
                    case 7:
                        allowance = checkFormikValue('propertyCity');
                        break;
                    case 8:
                        allowance = checkFormikValue('firstName') && checkFormikValue('lastName');
                        break;
                    case 9:
                        allowance = checkFormikValue('email');
                        break;
                    case 10:
                        allowance = checkFormikValue('phone');
                        break;
                }
            }
        }

        return allowance;
    }

    return (
        <div className={styles.actionButtonsWrapper}>
            {props.stepNumber > 1 &&
                <button
                    className={styles.actionButton}
                    type="button"
                    onClick={(): void => props.onPrevious()}
                >{`Previous`}</button>
            }
            {props.stepNumber !== props.totalSteps ?
                <button
                    className={styles.actionButton}
                    type="button"
                    onClick={(): void => props.onNext()}
                    disabled={!allowNextStep()}
                >{`Next`}</button> :
                <button
                    className={styles.actionButton}
                    type="button"
                    onClick={(): Promise<void> => props.formikProps.submitForm()}
                    disabled={!allowNextStep()}
                >
                    {!isLoading ? 'Submit' :
                        <FontAwesomeIcon
                            className={styles.loadingSpinner}
                            icon={faSpinner}
                            pulse
                        />
                    }
                </button>
            }
        </div>
    );
}

export default FormikActionButtons;
