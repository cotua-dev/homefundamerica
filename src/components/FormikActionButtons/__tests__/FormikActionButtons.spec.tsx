import React from 'react';
import { useFormik } from 'formik';
import { render, RenderResult } from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks';
import '@testing-library/jest-dom/extend-expect';
import { default as FormikActionButtons } from '..';
import { Props } from '../FormikActionButtons.interfaces';
import { FormikValues } from '../../FormikWizard/FormikWizard.interfaces';

describe('FormikActionButtons component', (): void => {
    test('renders', (): void => {
        const initialValues: FormikValues = {
            purchaseOrRefinance: null,
            propertyUse: null,
            propertyType: null,
            credit: null,
            borrowAmount: null,
            loanAmount: null,
            propertyValue: null,
            propertyCity: null,
            firstName: null,
            lastName: null,
            email: null,
            phone: null,
        };
        const { result } = renderHook(() => useFormik<FormikValues>({
            initialValues,
            onSubmit: () => {},
        }));

        act(() => {
            const props: Props = {
                canStep: false,
                stepNumber: 0,
                totalSteps: 0,
                formikProps: result.current,
                onNext: () => {},
                onPrevious: () => {},
            };

            const { container }: RenderResult = render(<FormikActionButtons {...props}/>);

            expect(container.firstChild).toMatchSnapshot();
        });
    });
});
