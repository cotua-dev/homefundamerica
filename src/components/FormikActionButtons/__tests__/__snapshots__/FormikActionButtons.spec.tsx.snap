// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormikActionButtons component renders 1`] = `
<div>
  <button
    disabled=""
    type="button"
  >
    Submit
  </button>
</div>
`;
